### ISSUE TITLE
A bug report: possible under-optimization issue

### ISSUE THREAD

**Issue by Anonymity20202** [emscripten-core/emscripten#17587]
The minimal code example:
```C
#include <stdio.h>
#include <stdint.h>

static volatile int32_t g_3 = 0x443EB606L;
static int32_t g_5 = 1L;

static uint32_t foo() {
  for (g_5 = 8; (g_5 >= 0); g_5 -= 1)
    for (g_3 = 0; g_3 < 9; g_3 += 1)
      ;
}
int main(int argc) {
  foo();
  printf("%d\n", g_3);
  printf("%d\n", g_5);
}
```

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.14 (4343cbec72b7db283ea3bda1adc6cb1811ae9a73)
clang version 15.0.0 (https://github.com/llvm/llvm-project 7effcbda49ba32991b8955821b8fdbd4f8f303e2)
Target: wasm32-unknown-emscripten
Thread model: posix
```

**Compilation option:**
`emcc -O3 -g -w test.c -o test.wasm -o test.js`

**Compiled WASM code (in text format):**
Extracted with [wabt](https://github.com/WebAssembly/wabt)
[test.txt](https://github.com/emscripten-core/emscripten/files/9275220/test.txt)

**Bug Description:**
The output of the WASM code is as expected (`9 -1`). However, even with the `O3` option, emcc seems failed to reach the best-optimized code. If compile the above C code with `gcc -O3` (version 11.2.0), all code in function `foo` is optimized away. But `emcc -O3` only inlines the `foo` function, most calculation and assignment in `foo` is still kept. 

As WASM emphasizes the performance on the web browsers, and as emcc is the most mature WASM compiler, although such an under-optimization issue does not incur severe problems (no functionality error), I think it is still worth a fixing?

**Comment by sbc100** [comment]
emscripten is based on clang and llvm so this is likely a difference between `gcc -O3` and `clang -O3`  .. can you confirm is using native clang gives you the same results as using emcc? (i.e. not completely optimizing away the function). 

I imagine that the volatile keyword is the thing that stop clang from optimizing here.  The keyword is saying that `g_5` could effectively change at any moment, and it could be changed by effects outside of the program itself.  I imagine that gcc and clang model the volatile keyword slightly differently here.

**Comment by Anonymity20202** [comment]
Thank you for your response! I have tried with native clang-12, and it gives the same results (not completely optimizing away):
```
objdump -S ./test_clang.out
...
int main(int argc) {
  401130:	50                   	push   %rax
  401131:	c7 05 f9 2e 00 00 08 	movl   $0x8,0x2ef9(%rip)        # 404034 <g_5>
  401138:	00 00 00 
    for (g_3 = 0; g_3 < 9; g_3 += 1)
  40113b:	c7 05 eb 2e 00 00 00 	movl   $0x0,0x2eeb(%rip)        # 404030 <g_3>
  401142:	00 00 00 
  401145:	8b 05 e5 2e 00 00    	mov    0x2ee5(%rip),%eax        # 404030 <g_3>
...
```
so it seems that this issue is caused by the difference between the optimization strategy of clang and gcc over the volatile keyword? Do you think such an issue deserves special attention? Or is it possible to fix it on the emcc side? like, improve the compiled WASM code since there seems to be no equivalent for the `volatile` keyword in WASM?

**Comment by juj** [comment]
> If compile the above C code with gcc -O3 (version 11.2.0), all code in function foo is optimized away.

This looks to me like a potential `gcc` compiler bug, might be worth reporting it to them. The C++ standard states https://en.cppreference.com/w/cpp/language/cv that

"
volatile object - an object whose type is volatile-qualified, or a subobject of a volatile object, or a mutable subobject of a const-volatile object. Every access (read or write operation, member function call, etc.) made through a glvalue expression of volatile-qualified type is treated as a visible side-effect for the [purposes of optimization](https://en.cppreference.com/w/cpp/language/as_if) (that is, within a single thread of execution, **volatile accesses cannot be optimized out** or reordered with another visible side effect that is [sequenced-before](https://en.cppreference.com/w/cpp/language/eval_order) or sequenced-after the volatile access.
"

**Comment by Anonymity20202** [comment]
Hi @juj and @sbc100, I observed a similar situation again. Here is the code:
```C
#include <stdio.h>
#include <stdint.h>
int32_t a, e, d; 
uint8_t g_64;
int32_t volatile g;
static int8_t foo(){
    for (a = 19; a >= -17; a--){
        for (d = 0; d <= 7; d++)
		    for (e = 0; e <= 7; e++){
                uint8_t *k = &g_64;
                *k = *k + 1;
                g = g;
            }
    }
}
void main(){
	foo();
	printf("g_64: %d\n", g_64);
}
```
While `g` is `volatile` and should not be optimized, the global variable `g_64` is optimizable. I compiled this program with gcc (9.4.0) and clang-12; both compilers will generate code similar to the following (in pseudocode):
```C
  do{
    g = g; // for 64 times
    --v3;
  }while ( v3 > -17 );
  d = 8;
  g_64 += 64;
  e = 8;
  a = -18;
```
However, emcc (3.1.14) generates code like this:
```C
  a = 0;
  do{
    g = g; // for 64 times
    --v3;
    d = 8;
    g_64 += 64;
    e = 8;
    a -= -1;
  }while ( v3 > -17 );

```
Although the result is correct, the WASM code contains unnecessary computation (under-optimization).

**Comment by juj** [comment]
Agreed. Cleaning up the test case to compile, and dropping the `printf` in favor of a `console.log` in order to make the compiler not pull in stdio for a simpler wasm-dis disassembly, building

`a.cpp`

```c++
#include <emscripten.h>
#include <stdio.h>
#include <stdint.h>

int32_t a, e, d;
uint8_t g_64;
int32_t volatile g;

void foo() {
  for (a = 19; a >= -17; a--) {
    for (d = 0; d <= 7; d++) {
      for (e = 0; e <= 7; e++) {
        uint8_t *k = &g_64;
        *k = *k + 1;
        g = g;
      }
    }
  }
}

int main() {
  foo();
  EM_ASM(console.log($0), g_64);
}
```
with `em++ a.cpp -o a.html -O3` does indeed produce an output that is not the cleanest that it could be.

The compiler fully unrolls the triple nested loop body, but still keeps a dummy `loop` block in the produced output, but always unconditionally `break`s out of that loop at the end of its first iteration (i.e. a `while(1) { ...; break; }` kind of structure).

If I remove the `volatile` keyword, then the code is cleaned up and the above inefficiency does not exist.

The optimizations that govern this type of behavior are produced by the general LLVM toolchain. There are no Emscripten-specific optimization passes or behavior at the general C/C++ IR level that would affect this behavior, to my knowledge. So this type of issue would go to the LLVM bug tracker - this is not something that Emscripten or Binaryen deal with.

(in some Wasm-specific cases, it might be something that the [Binaryen Wasm optimizer](https://github.com/WebAssembly/binaryen) might deal with, though this seems like a general LLVM IR type of scenario)

However, I would not be surprised if finding optimization inefficiencies around the `volatile` keyword is not likely to find motivation among LLVM devs. The compiler will quite likely tread very carefully whenever `volatile` is present, and many types of optimization passes will have their hands cuffed behind their backs because of the constraints that `volatile` brings.

> I compiled this program with gcc (9.4.0) and clang-12

In general it would be best not to compare with an older stable release of gcc or clang, but primarily test against the exact same release of clang that your installation of Emscripten uses. I.e. use the clang that is bundled with Emscripten to produce your native binaries. That way you'll have the identical native Clang compiler producing the native executables and the Emscripten WebAssembly code, and can do exact identical comparison from the same source IR compiler. That can help rule out if a particular behavior might be a clang-12 vs clang-14 issue for example.

Taking a step back, you seem to be dead set on a quest for being able to find some compiler codegen inefficiencies? Such an activity is certainly welcome, but especially so if you can pair such findings with a motivation where a possibly categorical amount of code will likely suffer from that kind of a lack of optimization. Codebases with `volatile` keyword are generally a bit exotic - especially so with Emscripten/WebAssembly, so I am uncertain if that will trigger a large interest or urgency from LLVM developers to address the issue.

To start, if you can find general codegen issues without the use of `volatile`, that would definitely step up the level of interest, since that would immediately be on the realm of affecting a lot of codebases. Or unless you have some specific interest to use `volatile`?
